### PR TITLE
[Work in Progress] Xata Workers (2/3)

### PR DESCRIPTION
**Pull request by gimenete**
# Xata Workers

Write code that queries Xata, and execute it safely client-side.

## Code structure

[Part 1](https://github.com/xataio/frontend-next/pull/1097): API Routes (frontend-next) [Private]
[Part 2](https://github.com/xataio/client-ts/pull/306): CLI and SDK (client-ts) [Public]
[Part 3](https://github.com/xataio/xata-workers/pull/1): Worker Router (xata-workers) [Private]

## Internal data modelling

- Worker: Compile of a collection of worker scripts
- Worker script: User code that can be executed
- Worker script executions: Requests done client side for a given worker script

## Usage concepts

1. Worker script code

```ts
const getBlogPosts = xataWorker("get-posts", async ({ xata }, id) => {
    return xata.db.posts.filter("blog.id", id).getAll();
});
```

2. Compile and publish step

```
xata workers upload
```

3. Local development mode

```
xata workers watch
```

## Implementation details

More details in the meta issue: https://github.com/xataio/frontend-next/issues/891

### CODE REVIEW

File: cli/src/commands/workers/upload.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import { Flags } from '@oclif/core';
+import fetch from 'node-fetch';
+import { z } from 'zod';
+import { BaseCommand } from '../../base.js';
+import { buildWatcher, compileWorkers, WorkerScript, workerScriptSchema } from '../../workers.js';
+
+const UPLOAD_ENDPOINT = 'https://xata-app-git-workers-xata.vercel.app/api/workers';
```
Comment (SferaDev) [inline cli/src/commands/workers/upload.ts:7]:
```suggestion
const UPLOAD_ENDPOINT = 'https://app.xata.io/api/workers';
```

File: cli/src/commands/workers/upload.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      action: async (path) => {
+        const compiledWorkers = await compileWorkers(path);
+        console.log(`Compiled ${compiledWorkers.length} workers`);
+
+        for (const worker of compiledWorkers) {
+          if (workers.has(worker.name)) {
+            this.error(`Worker ${worker.name} already exists. Worker names must be unique.`);
+          }
+
+          console.log('Saving worker', worker.name);
```
Comment (gimenete) [inline cli/src/commands/workers/upload.ts:47]:
```suggestion
          this.info('Saving worker', worker.name);
```

File: cli/src/commands/workers/upload.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    };
+
+    const response = await fetch(UPLOAD_ENDPOINT, {
+      method: 'POST',
+      headers: { Authorization: `Bearer ${profile.apiKey}` },
+      body: JSON.stringify(body)
+    });
+
+    const json = await response.json();
+    console.log(json);
```
Comment (gimenete) [inline cli/src/commands/workers/upload.ts:76]: Leftover?

File: cli/src/commands/workers/upload.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    const { id } = responseSchema.parse(json);
+
+    // TODO: Update codegen file and save
+    console.log(`Worker: ${id}`);
```
Comment (gimenete) [inline cli/src/commands/workers/upload.ts:81]:
```suggestion
    this.info(`Worker: ${id}`);
```

File: packages/client/src/util/request.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
+interface Body {
```
Comment (gimenete) [inline packages/client/src/util/request.ts:1]: Is this file copied from anywhere? If so, it'd be nice to leave a comment with the source so we can update in the future if we need to.

File: cli/src/commands/workers/watch.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    ]);
+
+    const miniflare = new Miniflare({
+      mounts: Object.fromEntries(mounts)
+    });
+
+    const server = await miniflare.createServer();
+
+    server.listen(watchPort, () => {
+      console.log(`Listening on port ${watchPort}`);
```
Comment (gimenete) [inline cli/src/commands/workers/watch.ts:55]:
```suggestion
      this.info(`Listening on port ${watchPort}`);
```

File: cli/src/commands/workers/watch.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      }
+    ]);
+
+    const miniflare = new Miniflare({
+      mounts: Object.fromEntries(mounts)
+    });
+
+    const server = await miniflare.createServer();
+
+    server.listen(watchPort, () => {
```
Comment (gimenete) [inline cli/src/commands/workers/watch.ts:54]: Can this port be ignored? Usually I prefer to ignore this parameter if possible so the system chooses a free port always and we don't have problems with running multiple watchers at the same time.

File: packages/client/src/workers/index.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+export function buildWorkerRunner<XataClient>(config: WorkerRunnerConfig) {
+  return function xataWorker<WorkerFunction extends (ctx: XataWorkerContext<XataClient>, ...args: any[]) => any>(
+    name: string,
+    _worker: WorkerFunction
+  ) {
+    return async (...args: RemoveFirst<Parameters<WorkerFunction>>): Promise<Awaited<ReturnType<typeof _worker>>> => {
+      const url =
+        process.env.NODE_ENV === 'development'
+          ? `http://localhost:64749/${name}`
```
Comment (gimenete) [inline packages/client/src/workers/index.ts:21]: I guess this won't work if the user specifies a different port in the CLI, right?

**Review by gimenete — COMMENTED** [review]
I cannot approve because I started this PR 😅 

I left a few comments, but nothing blocking. Can be addressed in follow-up PRs.

File: cli/src/commands/workers/upload.ts (hunk 3 above)
Comment (SferaDev) [inline cli/src/commands/workers/upload.ts:76]:
```suggestion
```

File: packages/client/src/workers/index.ts (hunk 8 above)
Comment (SferaDev) [reply packages/client/src/workers/index.ts:21]: Yeah, I'll remove the customization of the port for now and we revisit it

File: cli/src/commands/workers/watch.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      description: 'Exclude a glob pattern of files to compile'
+    }),
+    port: Flags.integer({
+      description: 'Port to use for the watcher'
+    })
+  };
+
+  async run(): Promise<void> {
+    const { flags } = await this.parse(WorkersCompile);
+    const watchPort = flags.port ?? 64749;
```
Comment (SferaDev) [inline cli/src/commands/workers/watch.ts:24]:
```suggestion
    // TODO: Allow customising port
    const watchPort = 64749;
```

File: cli/src/commands/workers/watch.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    ...this.databaseURLFlag,
+    include: Flags.string({
+      description: 'Include a glob pattern of files to compile'
+    }),
+    ignore: Flags.string({
+      description: 'Exclude a glob pattern of files to compile'
+    }),
+    port: Flags.integer({
+      description: 'Port to use for the watcher'
+    })
```
Comment (SferaDev) [inline cli/src/commands/workers/watch.ts:19]:
```suggestion
```

File: packages/client/src/util/request.ts (hunk 5 above)
Comment (SferaDev) [reply packages/client/src/util/request.ts:1]: Yeah, I took them from the typescript internal definitions (libdom)

Comment (SferaDev) [inline packages/client/src/util/request.ts:1]:
```suggestion
// Type definitions from libdom: https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts

interface Body {
```

**Review by SferaDev — APPROVED** [review]

File: cli/src/commands/workers/watch.ts (hunk 7 above)
Comment (SferaDev) [reply cli/src/commands/workers/watch.ts:52]: We need a known port for the client to connect, we can add an env variable to be set to customize it but I leave that as an improvement. Added a TODO comment in the code

**Review by SferaDev — APPROVED** [review]

**Pull request merged**